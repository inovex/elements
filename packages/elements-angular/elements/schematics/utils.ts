import { virtualFs, workspaces } from '@angular-devkit/core';
import { SchematicsException, Tree } from '@angular-devkit/schematics';

const CONFIG_PATH = 'angular.json';

function createHost(tree: Tree): workspaces.WorkspaceHost {
  return {
    async readFile(path: string): Promise<string> {
      const data = tree.read(path);
      if (!data) {
        throw new SchematicsException('File not found.');
      }
      return virtualFs.fileBufferToString(data);
    },
    async writeFile(path: string, data: string): Promise<void> {
      return tree.overwrite(path, data);
    },
    async isDirectory(path: string): Promise<boolean> {
      return !tree.exists(path) && tree.getDir(path).subfiles.length > 0;
    },
    async isFile(path: string): Promise<boolean> {
      return tree.exists(path);
    },
  };
}

export function readConfig(host: Tree) {
  const sourceText = host.read(CONFIG_PATH)?.toString('utf-8') || '';
  return JSON.parse(sourceText);
}

export function writeConfig(host: Tree, config: JSON) {
  host.overwrite(CONFIG_PATH, JSON.stringify(config, null, 2));
}

export function getAngularAppConfig(
  config: any,
  projectName: string
): any | never {
  if (!config.projects.hasOwnProperty(projectName)) {
    throw new SchematicsException(`Could not find project: ${projectName}`);
  }

  return config.projects[projectName];
}

export async function getProjectName(tree: Tree) {
  const host = createHost(tree);
  const { workspace } = await workspaces.readWorkspace('/', host);
  return workspace.extensions.defaultProject as string;
}
