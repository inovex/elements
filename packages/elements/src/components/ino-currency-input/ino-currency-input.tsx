import {
  Component,
  Element,
  Host,
  h,
  Prop,
  Event,
  EventEmitter,
  Watch,
} from '@stencil/core';
import { config } from '../config';

/**
 * @slot default - `<ino-input>` of `type="text"`
 */
@Component({
  tag: 'ino-currency-input',
  styleUrl: 'ino-currency-input.scss',
  shadow: false,
})
export class CurrencyInput {
  private readonly defaultLocale = 'de-DE';
  private inoInputEl?: HTMLInoInputElement;
  private hiddenInputEl?: HTMLInputElement;

  @Element() el!: HTMLElement;

  /**
   * A supported locale for currency number formatting. If not given, it uses the global config.
   * See https://developer.mozilla.org/de/docs/Web/JavaScript/Reference/Global_Objects/Intl#locales_argument
   */
  @Prop() currencyLocale?: string;

  /**
   * Numeric currency value
   */
  @Prop({ reflect: true }) value: number | string;
  @Watch('value')
  valueChanged(value: number | string) {
    if (!value || value === 'null') {
      this.hiddenValue = null;
      this.inputValue = null;
    }
    // Only set value if not from user input (has been set before)
    else if (value != this.hiddenInputEl?.value) {
      this.hiddenValue = value.toString();
      this.inputValue = this.toCurrencyString(value);
    }
  }

  /**
   * Emits when the user types something in.
   * Contains typed input in `event.detail`
   */
  @Event() valueChange!: EventEmitter<number>;

  componentDidLoad() {
    this.inoInputEl = this.el.querySelector('ino-input');
    this.inoInputEl?.setUserInputInterceptor((value) =>
      this.trimCurrencyString(value)
    );

    if (!this.inoInputEl) {
      console.error('currency input requires ino-input as child');
    }
    if (this.inoInputEl.type !== 'text') {
      console.error('currency input has to be of type "text"');
    }
    this.initialize();
  }

  connectedCallback() {
    // This is needed if the currency input is reattached to the dom (for instance in a popover)
    this.initialize();
  }

  disconnectedCallback() {
    this.inoInputEl?.removeEventListener('inoFocus', this.handleFocus);
    this.inoInputEl?.removeEventListener('inoBlur', this.handleBlur);
    this.inoInputEl?.removeEventListener('valueChange', this.handleValueChange);
  }

  private initialize() {
    // This is needed if the currency input is reattached to the dom (for instance in a popover)
    if (this.hiddenInputEl) {
      this.setupHiddenInput();
    }

    // Setup listener manually to only handle child events and catch non-bubbeling focus / blur.
    // @Listen would also trigger on valueChange of currency-input.
    this.inoInputEl?.addEventListener('inoFocus', this.handleFocus);
    this.inoInputEl?.addEventListener('inoBlur', this.handleBlur);
    this.inoInputEl?.addEventListener('valueChange', this.handleValueChange);
  }

  // -------
  // Event Handler

  private handleValueChange = (e: CustomEvent) => {
    let value = e.detail;
    // value is string => !"0" == true
    value = !value ? null : this.fromCurrencyString(value);

    // Set the hidden input value here and emit the event. This is not fully unmanaged,
    // but ensures that the value is always up-to-date and the value watcher doesn't format
    // the input (see valueChanged function).
    if (!Number.isNaN(value)) {
      this.hiddenValue = value;
      this.valueChange.emit(value);
    }
    e.stopPropagation();
  };

  private handleFocus = () => {
    this.inoInputEl.value = this.trimCurrencyString(this.inoInputEl.value);
  };

  private handleBlur = () => {
    const value = this.hiddenInputEl.value;
    if (value && value != 'null') {
      this.inoInputEl.value = this.toCurrencyString(value);
    }
  };

  // -------
  // Utils

  private set hiddenValue(value: string) {
    if (this.hiddenInputEl) {
      this.hiddenInputEl.value = value;
    }
  }
  private set inputValue(value: string) {
    if (this.inoInputEl) {
      this.inoInputEl.value = value;
    }
  }

  private setupHiddenInput() {
    this.hiddenValue = String(this.value);

    this.hiddenInputEl.name = this.inoInputEl.name;
    this.inoInputEl.name = null;

    this.hiddenInputEl.min = this.inoInputEl.min;
    this.hiddenInputEl.max = this.inoInputEl.max;
    this.hiddenInputEl.disabled = this.inoInputEl.disabled;
    this.hiddenInputEl.required = this.inoInputEl.required;

    this.inoInputEl.value = this.toCurrencyString(this.value);
  }

  // -----
  // Currency logic

  private getSeperator() {
    return this.toCurrencyString(0).charAt(1);
  }

  private toCurrencyString(value: string | number): string {
    if (!value  && value !== 0) return null;

    const locale = this.currencyLocale?.length > 0
      ? this.currencyLocale
      : config.get('currencyLocale', this.defaultLocale);

    const opts = { minimumFractionDigits: 2, maximumFractionDigits: 2 };
    return Intl.NumberFormat(locale, opts).format(Number(value));
  }

  private trimCurrencyString(value?: string): string {
    if (!value) {
      return value;
    }

    const leadingSymbol = this.extractLeadingSymbol(value);
    const separator = this.getSeperator();
    const numberParts = this.removeNonDecimalValues(value).split(separator);

    // Always keep the first part
    value = numberParts[0];
    // Decimals
    if (numberParts.length > 1) {
      // Only keep the first decimal occurence and only the given precision
      value += separator + numberParts[1].substring(0, 2);
    }
    return leadingSymbol + value;
  }

  private fromCurrencyString(value: string): number {
    const leadingSymbol = this.extractLeadingSymbol(value);
    const separator = this.getSeperator();
    const newValue = this.removeNonDecimalValues(value).replace(separator, '.');
    return Number(leadingSymbol + newValue);
  }

  private extractLeadingSymbol(value: string): string {
    const min = this.inoInputEl?.min;
    const requirePositive = min && Number(min) >= 0;
    return requirePositive ? '' : value.startsWith('-') ? '-' : '';
  }

  private removeNonDecimalValues(value: string) {
    const currencyRegex = new RegExp(`[^${this.getSeperator()}0-9]`, 'g');
    return value.replace(currencyRegex, '');
  }

  render() {
    return (
      <Host>
        <input
          type="hidden"
          ref={(el) => (this.hiddenInputEl = el)}
          value={this.value}
        />
        <slot></slot>
      </Host>
    );
  }
}
