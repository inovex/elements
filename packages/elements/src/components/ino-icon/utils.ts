import { getAssetPath } from '@stencil/core';

let CACHED_MAP: Map<string, string>;

export function getIconMap(): Map<string, string> {
	if (!CACHED_MAP) {
		const win = window as any;
		win.inoIcons = win.inoIcons || {};
		CACHED_MAP = win.inoIcons.map = win.inoIcons.map || new Map();
	}
	return CACHED_MAP;
}

export function addIcons(icons: { [name: string]: string }) {
	const map = getIconMap();
	Object.keys(icons).forEach((name) => map.set(name, icons[name]));
}

export const getSrc = (src: string | undefined) => {
	if (isStr(src)) {
		src = src.trim();
		if (isSrc(src)) {
			return src;
		}
	}
	return null;
};

export function getName(name: string | undefined) {
	if (name && !isSrc(name)) {
		return name;
	}
	return undefined;
}

export function getUrl(src: string | undefined, icon: string | undefined) {
	let url = getSrc(src);

	if (url) {
		return url;
	}
	url = getName(icon);
	if (url) {
		return getNamedUrl(url);
	}

	url = getSrc(icon);
	if (url) {
		return url;
	}

	return null;
}

export function getNamedUrl(name: string) {
	const url = getIconMap().get(name);
	if (url) {
		return url;
	}
	return getAssetPath(`./ino-icon/${name}.svg`);
}

export const isSrc = (str: string) => str.length > 0 && /(\/|\.)/.test(str);

export const isStr = (val: any): val is string => typeof val === 'string';

export const toLower = (val: string) => val.toLowerCase();
