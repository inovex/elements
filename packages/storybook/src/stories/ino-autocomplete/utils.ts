import { KeyValue } from '@inovex.de/elements';

export const AutoCompleteStoryOptions: KeyValue[] = [
  {
    key: 'hh',
    value: 'Hamburg',
  },
  {
    key: 'ber',
    value: 'Berlin',
  },
  {
    key: 'k',
    value: 'Karlsruhe',
  },
];
