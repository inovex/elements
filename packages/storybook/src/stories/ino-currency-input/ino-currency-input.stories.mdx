import { Meta, Description } from "@storybook/addon-docs";
import readme from "../../../../elements/src/components/ino-currency-input/readme.md";

{/* <Meta title="Input/ino-currency-input/Documentation" /> */}

<Description markdown={readme} />
