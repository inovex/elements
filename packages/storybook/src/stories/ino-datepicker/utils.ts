import '@inovex.de/elements';

export function registerInlineDatepickerHandler() {
  const group = document.querySelector('ino-radio-group') as HTMLInoRadioGroupElement;
  const datepicker = document.querySelector('.datepicker-group ino-datepicker') as HTMLInoDatepickerElement;

  const today = document.querySelector('.today');
  const thisWeek = document.querySelector('.thisWeek');
  const lastWeek = document.querySelector('.lastWeek');

  const todayClickHandler = () => {
    datepicker.value = '02.03.2020';
    datepicker.label = 'Am';
    group.value = 'at';
    datepicker.range = false;
  };

  const thisWeekClickHandler = () => {
    group.value = 'range';
    datepicker.label = 'Zeitraum';
    datepicker.range = true;
    datepicker.value = '02.03.2020 bis 08.03.2020';
  };

  const lastWeekClickHandler = () => {
    group.value = 'range';
    datepicker.label = 'Zeitraum';
    datepicker.range = true;
    datepicker.value = '24.02.2020 bis 01.03.2020';
  };

  const checkedChangeHandler = (e: CustomEvent<boolean>) => {
    const target = e.target as HTMLInoRadioElement;

    switch (target.value) {
      case 'at':
        datepicker.label = 'Am';
        datepicker.range = false;
        break;
      case 'after':
        datepicker.label = 'Ab';
        datepicker.range = false;
        if (group.value == 'range') {
          datepicker.value = datepicker.value.split(' - ').shift();
        }
        break;
      case 'before':
        datepicker.label = 'Bis';
        datepicker.range = false;
        if (group.value == 'range') {
          datepicker.value = datepicker.value.split(' - ').pop();
        }
        break;
      case 'range':
        datepicker.range = true;
        datepicker.label = 'Zeitraum';
        break;
    }
    group.value = target.value;
  };

  today.addEventListener('click', todayClickHandler);
  thisWeek.addEventListener('click', thisWeekClickHandler);
  lastWeek.addEventListener('click', lastWeekClickHandler);
  group.addEventListener('checkedChange', checkedChangeHandler);

  return () => {
    today.removeEventListener('click', todayClickHandler);
    thisWeek.removeEventListener('click', thisWeekClickHandler);
    lastWeek.removeEventListener('click', lastWeekClickHandler);
    group.removeEventListener('checkedChange', checkedChangeHandler);
  };
}
