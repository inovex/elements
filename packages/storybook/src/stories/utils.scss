/**
 * Applies styling to the container of a story.
 * The styling is encapsulated within this mixin because the ID may change in the future.
 */
@mixin story-container {
  #root-inner {
    @content;
  }
}
