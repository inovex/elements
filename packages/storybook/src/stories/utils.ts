import { html } from 'lit-html';
import ICONS from '../../../elements/src/components/ino-icon/icons';

export const maybeCreateStoryArgs = (story: any) => {
  if (!story.args) {
    story.args = {};
  }
  if (!story.argTypes) {
    story.argTypes = {};
  }
};

export const withIconControl = (
  story: any,
  propertyName: string,
  defaultValue?: string
) => {
  maybeCreateStoryArgs(story);
  story.args[propertyName] = defaultValue || '';
  story.argTypes[propertyName] = {
    control: {
      type: 'select',
    },
    options: getIcons(),
  };
};

export const withColorScheme = (
  story: any,
  propertyName: string,
  defaultValue?: string
) => {
  maybeCreateStoryArgs(story);
  story.args[propertyName] = defaultValue || '';
  story.argTypes[propertyName] = {
    control: {
      type: 'select',
    },
    options: [
      'primary',
      'secondary',
      'success',
      'warning',
      'error',
      'light',
      'dark',
    ],
  };
};

export const withSortDirection = (
  story: any,
  propertyName: string,
  defaultValue?: string
) => {
  maybeCreateStoryArgs(story);
  story.args[propertyName] = defaultValue || '';
  story.argTypes[propertyName] = {
    options: ['desc', 'asc'],
    control: {
      type: 'select',
    },
  };
};

export const getIcons = () => ICONS.filter((icon) => !icon.startsWith('_'));

export const defaultDecorator = (story, className?: string) => {
  return html` <div class="${className ?? ''}">${story()}</div> `;
};
