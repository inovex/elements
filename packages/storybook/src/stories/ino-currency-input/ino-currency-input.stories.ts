import { useEffect } from '@storybook/client-api';
import { Components } from '@inovex.de/elements';
import { html } from 'lit-html';
import { decorateStoryWithClass } from '../utils';
import './ino-currency-input.scss';
import { Story } from '@storybook/web-components';

export default {
  title: `Input/ino-currency-input`,
  component: 'ino-currency-input',
  decorators: [
    (story) => decorateStoryWithClass(story, 'story-ino-currency-input'),
    (story) => {
      useEffect(() => {
        const eventHandler = (e) => e.target.setAttribute('value', e.detail);
        const inputs = document.querySelectorAll('ino-currency-input');
        inputs.forEach((i) => i.addEventListener('valueChange', eventHandler));
        return () =>
          inputs.forEach((i) =>
            i.removeEventListener('valueChange', eventHandler)
          );
      });
      return story();
    },
  ],
  parameters: {
    actions: {
      handles: ['valueChange .customizable-input'],
    },
  },
};

export const Playground: Story<Components.InoCurrencyInput> = (args) => html`
  <ino-currency-input
    value="${args.value}"
    currency-locale="${args.currencyLocale}"
  >
    <ino-input
      class="customizable-input"
      id="customizable-input"
      autofocus="${args.autoFocus}"
      disabled="${args.disabled}"
      error="${args.error}"
      helper="${args.helper}"
      helper-persistent="${args.helperPersistent}"
      helper-validation="${args.helperValidation}"
      label="${args.label}"
      min="${args.min}"
      max="${args.max}"
      name="${args.name}"
      outline="${args.outline}"
      placeholder="${args.placeholder}"
      readonly="${args.readonly}"
      required="${args.required}"
      show-label-hint="${args.showLabelHint}"
      type="text"
      unit="${args.unit}"
    >
    </ino-input>
  </ino-currency-input>
`;
Playground.args = {
  autocomplete: false,
  autoFocus: false,
  disabled: false,
  error: false,
  helper: 'Helper message',
  helperPersistent: false,
  helperValidation: false,
  label: 'Currency Input',
  min: '',
  max: '',
  outlined: true,
  name: '',
  placeholder: '',
  readonly: false,
  required: false,
  showLabelHint: false,
  unit: '',
  value: '10.50',
  currencyLocale: '',
};
Playground.argTypes = {
  currencyLocale: {
    control: 'select',
    options: ['', 'de-DE', 'en-US'],
  },
};

export const DefaultValue = () => html`
  <ino-currency-input value="15.00">
    <ino-input type="text" label="Currency input"></ino-input>
  </ino-currency-input>
`;

export const Unit = () => html`
  <ino-currency-input value="15.00">
    <ino-input unit="€" label="Euro Input"></ino-input>
  </ino-currency-input>
`;

export const Locales = () => html`
  <ino-currency-input value="150215.99" currency-locale="de-DE">
    <ino-input autofocus unit="€" label="German locale"></ino-input>
  </ino-currency-input>

  <ino-currency-input value="150215.99" currency-locale="en-US">
    <ino-input unit="€" label="English locale"></ino-input>
  </ino-currency-input>
`;
