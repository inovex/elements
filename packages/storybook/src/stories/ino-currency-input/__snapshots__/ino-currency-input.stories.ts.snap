// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input/ino-currency-input Locales smoke-test 1`] = `
<div id="root-inner">
  <div class="story-ino-currency-input">
    <ino-currency-input value="15.00"
                        currency-locale="en-EN"
                        class="hydrated"
    >
      <input type="hidden"
             value="15.00"
             name="undefined"
             min="undefined"
             max="undefined"
      >
      <ino-input class="customizable-input hydrated"
                 id="customizable-input"
                 label="Default Input"
                 outline="true"
                 unit="€"
                 tabindex="-1"
      >
        <label class="ino-input__composer mdc-text-field mdc-text-field--outlined mdc-text-field--with-trailing-icon mdc-text-field--label-floating">
          <ino-label class="ino-label--outlined hydrated">
            <div class="mdc-notched-outline mdc-notched-outline--upgraded mdc-notched-outline--notched">
              <div class="mdc-notched-outline__leading">
              </div>
              <div class="mdc-notched-outline__notch"
                   style="width: 79.25px;"
              >
                <label class="mdc-floating-label mdc-floating-label--float-above"
                       style="transition-duration: 0s;"
                >
                  Default Input
                </label>
              </div>
              <div class="mdc-notched-outline__trailing">
              </div>
            </div>
          </ino-label>
          <input class="mdc-text-field__input"
                 step="1"
                 type="text"
          >
          <span class="mdc-text-field__affix mdc-text-field__affix--suffix">
            €
          </span>
        </label>
        <div class="mdc-text-field-helper-line">
        </div>
      </ino-input>
    </ino-currency-input>
  </div>
</div>
`;

exports[`Input/ino-currency-input Playground smoke-test 1`] = `
<div id="root-inner">
  <div class="story-ino-currency-input">
    <ino-currency-input value="15.00"
                        currency-locale="de-DE"
                        class="hydrated"
    >
      <input type="hidden"
             value="15.00"
             name="undefined"
             min="undefined"
             max="undefined"
      >
      <ino-input class="customizable-input hydrated"
                 id="customizable-input"
                 label="Default Input"
                 outline="true"
                 unit="€"
                 tabindex="-1"
      >
        <label class="ino-input__composer mdc-text-field mdc-text-field--outlined mdc-text-field--with-trailing-icon mdc-text-field--label-floating">
          <ino-label class="ino-label--outlined hydrated">
            <div class="mdc-notched-outline mdc-notched-outline--upgraded mdc-notched-outline--notched">
              <div class="mdc-notched-outline__leading">
              </div>
              <div class="mdc-notched-outline__notch"
                   style="width: 79.25px;"
              >
                <label class="mdc-floating-label mdc-floating-label--float-above"
                       style
                >
                  Default Input
                </label>
              </div>
              <div class="mdc-notched-outline__trailing">
              </div>
            </div>
          </ino-label>
          <input class="mdc-text-field__input"
                 step="1"
                 type="text"
          >
          <span class="mdc-text-field__affix mdc-text-field__affix--suffix">
            €
          </span>
        </label>
        <div class="mdc-text-field-helper-line">
        </div>
      </ino-input>
    </ino-currency-input>
  </div>
</div>
`;
